include/openthread/tcp.h

```diff
@@ -0,0 +1,616 @@
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ * @brief
+ *  This file defines the OpenThread TCP API.
+ *
+ */
+
+#ifndef OPENTHREAD_TCP_H_
+#define OPENTHREAD_TCP_H_
+
+#include <openthread/ip6.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @addtogroup api-tcp
+ *
+ * @brief
+ *   This module includes functions that control TCP communication.
+ *
+ * @{
+ *
+ */
+
+/**
+ * This a linked buffer structure for use with TCP.
+ *
+ * A single @p otLinkedBuffer structure references an array of bytes in memory,
+ * via @p mData and @p mLength. The @p mNext field is used to form a chain of
+ * @p otLinkedBuffer structures.
+ *
+ */
+typedef struct otLinkedBuffer
+{
+    struct otLinkedBuffer *mNext;  ///< Pointer to the next linked buffer in this chain, or NULL if this is the end of the chain
+    const uint8_t *mData;          ///< Pointer to data referenced by this linked buffer
+    uint16_t mLength;              ///< Length of this linked buffer
+} otLinkedBuffer;
+
+struct otTcpEndpoint;
+typedef struct otTcpEndpoint otTcpEndpoint;
+
+/**
+ * This callback informs the application that the TCP 3-way handshake is
+ * complete and that the connection is now established.
+ */
+typedef void (*otTcpEstablished)(void *aContext, otTcpEndpoint *aEndpoint);
+
+/**
+ * This callback informs the application that data in the provided
+ * @p otLinkedBuffers have been acknowledged by the connection peer and that
+ * the relevant @p otLinkedBuffers and the data they contain can be reclaimed
+ * by the application. The @p otLinkedBuffers are guaranteed to be identical to
+ * those passed in to TCP via @p otTcpSendByReference.
+ */
+typedef void (*otTcpSendDone)(void *aContext, otTcpEndpoint *aEndpoint, otLinkedBuffer *aData);
+
+/**
+ * This callback informs that the first @p aNumBytes in the send buffer have
+ * been acknowledged by the connection peer and that their underlying memory
+ * can be reclaimed by the application.
+ *
+ * This callback is not necessary for correct TCP operation. Most applications
+ * can just rely on the @p otTcpSendDone callback. If an application wants to
+ * be able to receive fine-grained feedback as memory in the send buffer
+ * becomes available again (instead of waiting for an entire linked buffer's
+ * worth of data becomes available), it can register this callback.
+ */
+typedef void (*otTcpBytesAcked)(void *aContext, otTcpEndpoint *aEndpoint, size_t aNumBytes);
+
+/**
+ * This callback informs the application that if data is added to the send
```

Comment:
nit: have brief in separate block above

Reply:
Done. I'm not 100% what you meant by "in separate block"; can you please check that what I've done is acceptable?